### DISCUSSION TITLE
Why is there no List/Dict/etc Model?

### DISCUSSION THREAD

**Discussion by mimre25** [pydantic/pydantic#6425] [Question] [answered]
Hi,

I've found myself often using lists or dicts as `root` (/`__root__`) attribute and resorting to what is described in the [documentation](https://docs.pydantic.dev/latest/usage/models/#rootmodel-and-custom-root-types) for making models iterable:
```py
T = TypeVar("T")
class ListModel<T>(RootModel):
    root: list[T]

    def __iter__(self):
        return iter(self.root)

    def __getitem__(self, item):
        return self.root[item]

...

class MyModel(ListModel[int]):
    pass

...

foo = MyModel([1,2,3,4])
for i in foo:
    ...
```

I feel like it was a conscious decision that such a model does not exist in `.

Could you maybe share some insights as to why this might be a bad idea/why it isn't included in `pydantic`?

**Comment by adriangb** [comment, marked as answer]
Is there any particular reason why you’re using RootModel? Can you just use List[T] directly, perhaps with TypeAdapter?

**Reply by mimre25** [reply]
Think of more complex scenarios:

```py
class Foo(BaseModel):
    ...

class Bar(BaseModel):
   ...

class Baz(RootModel):
    root: list[tuple(Foo, Bar)]
```

it's kinda handy for validation and json en- & de-coding.

**Reply by adriangb** [reply]
I don’t get it. Why not just `TypeAdapter(list[tuple[Foo, Bar]]).validate_json(…)`?

**Reply by mimre25** [reply]
Is `TypeAdapter` something new in V2? (I haven't used V2 yet, only V1.*)

If so, does it support `parse_json` and `parse_raw`?

**Reply by adriangb** [reply]
Yes, see https://docs.pydantic.dev/latest/migration/#introduction-of-typeadapter

**Reply by mimre25** [reply]
Ah cool stuff. 
I haven't read much into it and just played around a bit, but generally does what I need.

Just one thing I can think of regarding ergonomics that is lacking:

```py

class Foo(BaseModel):
    ...

x = TypeAdapter(list[int]).parse_json("[1,2,3]")
y = Foo(...)

models = [x,y]

for m in models:
    m.model_dump_json()

# or, alternatively
T = TypeVar("T", bound="BaseModel")
def serializer(some_model: T) -> str:
    return some_model.model_dump_json()
```

Another example could a a deserializer that just tries different models a la:

```py
REQUEST_MODELS = [...]

def deserialize(raw_data: bytes | str) -> ...:
    for model in REQUEST_MODELS:
        try:
            return model.model_validate_json()
        except:
            pass
```

Here doing the way around to add a `TypeAdapter` for some of the models seems not really ergonomic. 

That said, I can see how these cases are rather niche, and not necessary to be supported from pydantic out of the box :slightly_smiling_face: 

Generally speaking, if I encounter such a case, would there be anything bad in using the `ListModel` I showed in my original post?

**Reply by adriangb** [reply]
No, I guess you can do that if you want. You can also do `ListModel = RootModel[List[…]]`. But I still think you could just use TypeAdapter[tuple[list[int], Foo]) or create a list[TypeAdapter[…]], and are perhaps searching for a problem that fits the solution.

**Reply by mimre25** [reply]
> are perhaps searching for a problem that fits the solution.

I have to agree to some degree. Thank you for all the answer :raised_hands:

**Comment by acampove** [comment]
@adriangb I do not think TypeAdapter is enough. This type adapter idea just adds validation, but as @adriangb mentioned, there is serialization and deserialization functionality that we would need and cannot be added this way. We would need an actual model for that. For instance I am trying to build config classes that load the data from a file. For that I would do something like:

```python
    @classmethod
    def from_package(cls, file_path : str, package : str) -> Self:
        '''
        Parameters
        ---------------
        file_path: Path to config, wrt package, e.g. local/config.yaml
        '''
        data = gut.load_data(package = package, fpath = file_path)

        return cls(**data)
```

but I need something to hold the `from_package` method.
